### PR TITLE
Improved joinURLs so that two qualified URLs can be joined.

### PR DESCRIPTION
**Pull request by keattang**
Previously, joining `http://www.domain.com/some/path` with `http://www.domain.com/another/path` would result in `http://www.domain.com/another/path`. Now it results in `http://www.domain.com/some/path/another/path` which is more intuitive and useful.

Also tweaked `CesiumTerrainProvider` so that the tile URL templates can be overridden from the options.

### CODE REVIEW

**Comment by hpinkos** [comment]
Thanks @keattang!  This change seems fine to me, does anyone else have an opinion?

File: Source/Core/joinUrls.js [hunk 1]
```diff
@@ -80,6 +80,10 @@ define([
                 url += first.path + second.path;
             }
         } else {
+            url += first.path.replace(/^\/?/g, '');
+
+            if (appendSlash && second.path !== '') {url = url.replace(/\/?$/, '/');}
```
Comment (hpinkos) [inline Source/Core/joinUrls.js]: Please put these on separate lines to follow our code styling convensions

File: Specs/Core/joinUrlsSpec.js [hunk 2]
```diff
@@ -108,7 +108,7 @@ defineSuite([
 
     it('appends qualfied url with path correctly to qualified url with path', function() {
         var result = joinUrls(qualifiedUrlWithPath, qualifiedUrlWithPath);
-        expect(result).toEqual(qualifiedUrlWithPath);
+        expect(result).toEqual(qualifiedUrlWithPath+absolutePath);
```
Comment (hpinkos) [inline Specs/Core/joinUrlsSpec.js]: This is really nit picky, but can you please add spaces around the `+`?  Thanks!

**Comment by keattang** [comment]
Done :)

**Comment by kring** [comment]
@keattang and @hpinkos I'm not really sure the new logic makes sense to me.  In my mind, `joinUrls` is similar to .NET's [Path.Combine](https://msdn.microsoft.com/en-us/library/fyy7a5kt(v=vs.110).aspx) or Apache Commons' [FilenameUtils.concat](https://commons.apache.org/proper/commons-io/javadocs/api-1.4/org/apache/commons/io/FilenameUtils.html#concat(java.lang.String, java.lang.String)), both of which will return the second path and ignore the first if the second is absolute.  I'm not really sure why you would want joinUrls to work in the way you've described.  If `http://www.domain.com/another/path` is a valid URL, there's no reason to expect that `http://www.domain.com/some/path/another/path` is a URL that makes any sense at all.

**Comment by keattang** [comment]
Hey @kring I made the changes so that I could feed `CesiumTerrainProvider` a url that looks like `https://cloudfront.com/some/folder/` and a list of URL templates that look like `https://cloudfront1.com/{z}/{x}/{y}.terrain`, `https://cloudfront2.com/{z}/{x}/{y}.terrain`, etc. and get URLs out that look like `https://cloudfront1.com/some/folder/{z}/{x}/{y}.terrain`. 

It's a bit hacky but the changes to joinUrls to support it seemed to make sense. However if you think they're not appropriate I'm happy figure out another way to do it. Either way though I'll probably need the ability to override the `tiles` option so I might make that into a separate PR.

**Comment by kring** [comment]
Sorry, I still don't understand where such URLs would come from.  If you'll be overriding the tiles option anyway, why not just provide the correct absolute path in the first place?

**Comment by mramato** [comment]
@kring What's the status here, are we just closing this?

**Comment by hpinkos** [comment]
Yeah, thinking about this more, I don't think it makes sense for how we intended the `joinUrls` function to work.  I'm going to close this PR.

**Pull request closed without merging**